pkg/runtime.v2/core/trainingruntime.go

```diff
@@ -0,0 +1,121 @@
+	fwkplugins "github.com/kubeflow/training-operator/pkg/runtime.v2/framework/plugins"
+	idxer "github.com/kubeflow/training-operator/pkg/runtime.v2/indexer"
+)
+
+var (
+	errorNotFoundSpecifiedTrainingRuntime = errors.New("not found TrainingRuntime specified in TrainJob")
+)
+
+type TrainingRuntime struct {
+	framework *fwkcore.Framework
+	client    client.Client
+	scheme    *apiruntime.Scheme
+}
+
+var TrainingRuntimeGroupKind = schema.GroupKind{
+	Group: kubeflowv2.GroupVersion.Group,
+	Kind:  "TrainingRuntime",
+}.String()
+
+var _ runtime.Runtime = (*TrainingRuntime)(nil)
+
+var trainingRuntimeFactory *TrainingRuntime
+
+func NewTrainingRuntime(ctx context.Context, c client.Client, indexer client.FieldIndexer) (runtime.Runtime, error) {
+	if err := indexer.IndexField(ctx, &kubeflowv2.TrainJob{}, idxer.TrainJobTrainingRuntimeRefKey, idxer.IndexTrainJobTrainingRuntimes); err != nil {
+		return nil, fmt.Errorf("setting index on TrainingRuntime and ClusterTrainigRuntime for TrainJob: %w", err)
+	}
+	fwk, err := fwkcore.New(ctx, c, fwkplugins.NewRegistry(), indexer)
+	if err != nil {
+		return nil, err
+	}
+	trainingRuntimeFactory = &TrainingRuntime{
+		framework: fwk,
+		client:    c,
+		scheme:    c.Scheme(),
+	}
+	return trainingRuntimeFactory, nil
+}
+
+func (r *TrainingRuntime) NewObjects(ctx context.Context, trainJob *kubeflowv2.TrainJob) ([]client.Object, error) {
+	var trainingRuntime kubeflowv2.TrainingRuntime
+	err := r.client.Get(ctx, client.ObjectKey{Namespace: trainJob.Namespace, Name: trainJob.Spec.TrainingRuntimeRef.Name}, &trainingRuntime)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", errorNotFoundSpecifiedTrainingRuntime, err)
+	}
+	return r.buildObjects(ctx, trainJob, trainingRuntime.Spec.Template, trainingRuntime.Spec.MLPolicy, trainingRuntime.Spec.PodGroupPolicy)
+}
+
+func (r *TrainingRuntime) buildObjects(ctx context.Context, trainJob *kubeflowv2.TrainJob, jobSetTemplateSpec kubeflowv2.JobSetTemplateSpec,
+	mlPolicy *kubeflowv2.MLPolicy, podGroupPolicy *kubeflowv2.PodGroupPolicy) ([]client.Object, error) {
+	opts := []runtime.InfoOption{
+		runtime.WithLabels(jobSetTemplateSpec.Labels),
+		runtime.WithAnnotations(jobSetTemplateSpec.Annotations),
+		runtime.WithMLPolicy(mlPolicy),
+		runtime.WithPodGroupPolicy(podGroupPolicy),
+	}
+	for idx, rJob := range jobSetTemplateSpec.Spec.ReplicatedJobs {
+		if rJob.Replicas == 0 {
+			jobSetTemplateSpec.Spec.ReplicatedJobs[idx].Replicas = 1
+		}
+		replicas := jobSetTemplateSpec.Spec.ReplicatedJobs[idx].Replicas * ptr.Deref(rJob.Template.Spec.Completions, 1)
+		opts = append(opts, runtime.WithPodSpecReplicas(rJob.Name, replicas, rJob.Template.Spec.Template.Spec))
+	}
+	info := runtime.NewInfo(&jobsetv1alpha2.JobSet{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: jobsetv1alpha2.SchemeGroupVersion.String(),
+			Kind:       "JobSet",
+		},
+		Spec: *jobSetTemplateSpec.Spec.DeepCopy(),
+	}, opts...)
+
+	if err := r.framework.RunEnforceMLPolicyPlugins(info); err != nil {
```

Comment:
How we can make sure only one of MLPolicy will be used ?
E.g. user can set one of Torch or MPI policy.

Reply:
>Theoretically, we can perform multiple MLPolicy against a single TrainJob.

In that case, how we will ensure that some MLPolicies can be enforced only once and some of them not ?

Reply:
> In that case, how we will ensure that some MLPolicies can be enforced only once and some of them not ?

Do you assume the case of user defined plugins? In that case, the responsibility depends on the users.
They need to understand our kubeflow upstream plugins how the upstream ones should work for each extension points.

Reply:
I don't really see a use-case when users might have two MLPolicies they want to enforce.
@tenzen-y Can you give me any examples ?

Reply:
For example, they can add a custom plugin like `CompanyAModelParam` for PyTorch training.
In that case, they can keep using the upstream basic `.spec.mlPolicy.torch.numProcPerNode` policy, but they can add arbitrary parameters.

Reply:
Sounds good.